styled_text/src/attributed_text.rs

```diff
@@ -0,0 +1,243 @@
+// Copyright 2025 the Styled Text Authors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+use alloc::vec::Vec;
+use core::fmt::Debug;
+use core::ops::Range;
+
+use crate::{EditBehavior, EditableTextStorage, SpanEditAction, TextStorage};
+
+/// The errors that might happen as a result of [applying] an attribute.
+///
+/// [applying]: AttributedText::apply_attribute
+///
+/// TODO: impl Error for this.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum ApplyAttributeError {
+    /// The bounds given were invalid.
+    ///
+    /// TODO: Store some data about this here.
+    InvalidBounds,
+}
+
+/// A block of text with attributes applied to ranges within the text.
+#[derive(Debug)]
+pub struct AttributedText<T: Debug + TextStorage, Attr: Debug> {
+    text: T,
+    attributes: Vec<(Range<usize>, Attr)>,
+}
+
+impl<T: Debug + TextStorage, Attr: Debug> AttributedText<T, Attr> {
+    /// Create an `AttributedText` with no attributes applied.
+    pub fn new(text: T) -> Self {
+        Self {
+            text,
+            attributes: Vec::default(),
+        }
+    }
+
+    /// Apply an `attribute` to a `range` within the text.
+    pub fn apply_attribute(
+        &mut self,
+        range: Range<usize>,
+        attribute: Attr,
+    ) -> Result<(), ApplyAttributeError> {
+        let text_len = self.text.len();
+        if range.start > text_len || range.end > text_len {
+            return Err(ApplyAttributeError::InvalidBounds);
+        }
```

Comment:
It may be good to enforce consistency by ensuring attribute ranges span entire characters. This should only have to be checked when first applying an attribute, the ranges should then stay consistent.

```suggestion
        if !self.text.is_char_boundary(range.start) || !self.text.is_char_boundary(range.end)
        {
            return Err(ApplyAttributeError::InvalidBounds);
        }
```

(This keeps the within-bounds semantics as `text.is_char_boundary(idx)` is `true` for `idx == text.len()`, `false` for `idx > text.len()`.)